Source/messaging/ConsoleStreamRedirect.h

```diff
@@ -0,0 +1,498 @@
+							::CloseHandle(_handle);
+						}
+						Reader::Flush();
+					}
+					return (_index == Core::IResource::INVALID);
+				}
+
+			private:
+				bool CreateOverlappedPipe(HANDLE& readPipe, int& writePipe)
+				{
+					static volatile uint32_t sequenceNumber = 0;
+					uint32_t newIndex = Core::InterlockedIncrement(sequenceNumber);
+
+					HANDLE pipeHandle, pipeOutput;
+					TCHAR PipeNameBuffer[MAX_PATH];
+
+					sprintf(PipeNameBuffer,
+						_T("\\\\.\\Pipe\\ThunderRedirectPipe.%08x.%08x"), Core::ProcessInfo().Id(), newIndex);
+
+					pipeHandle = CreateNamedPipeA(
+						PipeNameBuffer,
+						PIPE_ACCESS_INBOUND | FILE_FLAG_OVERLAPPED,
+						PIPE_TYPE_BYTE | PIPE_WAIT,
+						1,
+						4096,
+						4096,
+						120 * 1000,
+						nullptr
+					);
+
+					if (pipeHandle != INVALID_HANDLE_VALUE) {
+						pipeOutput = CreateFileA(
+							PipeNameBuffer,
+							GENERIC_WRITE,
+							0,
+							nullptr,
+							OPEN_EXISTING,
+							FILE_ATTRIBUTE_NORMAL,
+							nullptr);
+
+						if (pipeOutput == INVALID_HANDLE_VALUE) {
+							::CloseHandle(pipeHandle);
+						}
+						else {
+							writePipe = _open_osfhandle(reinterpret_cast<intptr_t>(pipeOutput), 0);
+							if (writePipe == -1) {
+								::CloseHandle(pipeHandle);
+								::CloseHandle(pipeOutput);
+							}
+							else {
+								readPipe = pipeHandle;
+								return (true);
+							}
+						}
+					}
+					return(false);
+				}
+				void Read(OVERLAPPED& event) {
+					::ReadFile(_handle, Reader::Buffer(), Reader::Length(), nullptr, &event);
+				}
+				void Cancel(OVERLAPPED& event) {
+					// Maybe some data was still flushed, first flush that...
+					DWORD bytesRead = 0;
+
+					if ((::GetOverlappedResult(_handle, &event, &bytesRead, FALSE) != 0) && (bytesRead > 0)) {
+						Reader::ProcessBuffer(bytesRead);
+					}
+
+					::CancelIoEx(_handle, &event);
+				}
+				void Process(OVERLAPPED& event) {
+					DWORD bytesRead = 0;
+
+					if ((::GetOverlappedResult(_handle, &event, &bytesRead, FALSE) != 0) && (bytesRead > 0)) {
+						Reader::ProcessBuffer(bytesRead);
+						Read(event);
+					}
+				}
+
+			private:
+				Core::IResource::handle _index;
+				Core::IResource::handle _copy;
+				HANDLE _handle;
+			};
+
+			// friend class Core::SingletonType<ReaderImplementation::ResourceMonitor>;
+#else
+			class ReaderImplementation : public Core::IResource, public Reader {
+			public:
+				ReaderImplementation() = delete;
+				ReaderImplementation(ReaderImplementation&&) = delete;
+				ReaderImplementation(const ReaderImplementation&) = delete;
+				ReaderImplementation& operator=(ReaderImplementation&&) = delete;
+				ReaderImplementation& operator=(const ReaderImplementation&) = delete;
+
+				ReaderImplementation(ParentClass& parent)
+					: Reader(parent)
+					, _index(Core::IResource::INVALID)
+					, _copy(Core::IResource::INVALID)
+					, _handle(Core::IResource::INVALID) {
+				}
+				~ReaderImplementation() override {
+					Close();
+				}
+
+			public:
+
```

Comment:
Are these methods below not formatted on purpose? :smile:

Reply:
Yes that is on purpose, here you can see that the Linux system has a much cleaner structure than the mess they created in Windows. 
The open and close are different on Windows and Linux hence why there are 2 implementations. In linux the system is consotent everything behaves as a file descriptor. Sockets, Files and SerialPorts can be treated in Linux as a filedescriptor with a-synchronous IO calls, that are all monitored by the the *Core*::ResourceMonitor.
On Windows it is a mess and every sub component of Windows has a slightly different API, hence why here (and in for example the SerialPort handling here: https://github.com/rdkcentral/Thunder/blob/master/Source/core/SerialPort.cpp#L82-L300) we use a bit of a different approach to handle.
I do think we could align it by extending the Core::IResource so that all a-synchronous IO on Windows is also done by the ResourceMonitor but as Windows is officially is not supported it is something we need to do if time permits.